### ISSUE TITLE
ParenPad: bug in lambda support

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3329]
$ cat LambdaParams.java

``` java
class LambdaParams
{
    {
        java.util.function.Consumer c = ( o ) -> { o.toString() }; // 2 VIOLATIONS -- want no space after opening '(' and no space before closing ')' in lambda expression

        java.util.stream.Stream.of().forEach(( o ) -> o.toString()); // 2 VIOLATIONS -- want no space after opening '(' and no space before closing ')' in lambda expression

        java.util.stream.Stream.of().forEach(( Object o ) -> o.toString()); // 2 VIOLATIONS -- want no space after opening '(' and no space before closing ')' in lambda expression
    }

    void someMethod( String param ) // 2 VIOLATIONS -- want no space after opening '(' and no space before closing ')' in method def
    {
    }
}
```

$ cat config.xml

``` xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="ParenPad">
            <message key="ws.followed" value="Whitespace after opening ''{0}'' should be removed" />
            <message key="ws.preceded" value="Whitespace before closing ''{0}'' should be removed" />
        </module>
    </module>
</module>
```

$ java -jar checkstyle-7.0-all.jar -c config.xml LambdaParams.java
Starting audit...
[ERROR] C:\Users\U0076777\LambdaParams.java:6:47: Whitespace after opening '(' should be removed [ParenPad]
[ERROR] C:\Users\U0076777\LambdaParams.java:6:49: Whitespace before closing ')' should be removed [ParenPad]
[ERROR] C:\Users\U0076777\LambdaParams.java:8:47: Whitespace after opening '(' should be removed [ParenPad]
[ERROR] C:\Users\U0076777\LambdaParams.java:8:56: Whitespace before closing ')' should be removed [ParenPad]
[ERROR] C:\Users\U0076777\LambdaParams.java:11:21: Whitespace after opening '(' should be removed [ParenPad]
[ERROR] C:\Users\U0076777\LambdaParams.java:11:34: Whitespace before closing ')' should be removed [ParenPad]
Audit done.
Checkstyle ends with 6 errors.

---

I'm not sure what is different about the assignment statement, but that lambda is not producing violations for the ParenPad check.  There should be a total of 8 violations in the sample source.
So line 4 should produce 2 violations.
##

**Comment by soon** [comment]
Working on this issue. 

In the line `java.util.function.Consumer c = ( o ) -> { o.toString() };` lambda is parsed as a `LAMBDA` token, but `ParenPadCheck` module does not accept lambdas and the checker skips the lambda.

In the line `java.util.stream.Stream.of().forEach(( o ) -> o.toString());` lambda is a part of `ELIST` of `METHOD_CALL` and therefore processed as an expression.

My first impression was to add a `TokenType.LAMBDA` into a list of acceptable tokens, but I'm not sure if this is a correct solution.

**Comment by rnveach** [comment]
> My first impression was to add a TokenType.LAMBDA into a list of acceptable tokens

This was mine as well. There is no other way for the check to examine this token unless we tell it to visit it.
`visitToken` might have to change depending on if `lambda` needs special handling than the default.

**Comment by soon** [comment]
@rnveach Suppose we're excluding `LAMBDA` from the tokens list via config. Should `java.util.stream.Stream.of().forEach(( o ) -> o.toString());` raise a violation it that case?

**Comment by romani** [comment]
> raise a violation it that case?

no.
http://checkstyle.sourceforge.net/config_whitespace.html#ParenPad
See example. The same as if we remove "CTOR_CALL" from config, violations is disappear.

It would be super great if you put few examples of code+config to xdoc file (html site is generated from xdoc files) to show user how exactly config force code to looks like. Users like examples.

**Comment by romani** [comment]
fix is merged.
